### PR TITLE
Default Parameters

### PR DESCRIPTION
**Pull request by rwdougla**

### CODE REVIEW

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [inline functions/defaulted_parameters.md:52]: Are you thinking here of something like `void foo(const std::string& s = "hello world")`? Either way, I think you should be. :)  C++ treats the default argument value here _differently_ from how Python famously treats it.

However, note that this is a caveat about _defining_ functions, not _calling_ them.

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [inline functions/defaulted_parameters.md:66]: `Implementing a function with default arguments`

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [inline functions/defaulted_parameters.md:80]: Utter nit: `rightmost`, `leftward` (in American English at least)

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [inline functions/defaulted_parameters.md:86]: I don't understand what this sentence is getting at. (I assume `consumer` should be `caller`, but that doesn't help me.)

I would add here: `Similarities and differences between default function arguments and default template arguments.`

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [inline functions/defaulted_parameters.md:35]: `Using a function with default arguments`

And honestly, is there anything at all to teach here? Calling a function with or without default arguments looks and works the same way: `foo(some, args, here)`. I think "Defining a function with default arguments" should be included in the "Basic" part, and maybe there's just no "Advanced" part at all (except for teaching [the vast number of caveats that militate against ever using default arguments](https://quuxplusone.github.io/blog/2020/04/18/default-function-arguments-are-the-devil/)).

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [inline functions/defaulted_parameters.md:82]: When (if ever) to use default function arguments instead of a plain old overload set.

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [inline functions/defaulted_parameters.md:6]: Nit: Could you break this long line into 75- or 100-character lines?

`consumer` should presumably be `caller` (twice).

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [inline functions/defaulted_parameters.md:29]: This raises an interesting discussion point/caveat: At first I thought you were suggesting that one use-case for default arguments is if some third-party library provides the existing signature `void foo(int x, int y)`, the student might consider overriding it in their own code with `void foo(int x, int y = 42)` to save on typing without breaking ABI.
But then I decided that you must mean that if _the student_ has existing code that does `bool bar(int x)`, they can replace all declarations of that function with `bool bar(int x, int y = 42)` to preserve API... but notice that that change breaks ABI instead! (Also [breaks API for anyone who used `bar` as a function pointer](https://quuxplusone.github.io/blog/2020/04/18/default-function-arguments-are-the-devil/#surprising-function-pointer-type) — e.g. `std::find_if(first, last, bar)` — but that's orthogonal.)

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md:6]: I can certainly do either. Do we have a preferred convention?

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md:29]: From a practical standpoint, I often find default parameters come from having a function, `foo(int x)` where the maintainer needs to add a parameter to become `foo(int x, int y)` but to avoid breaking all existing calls, they give it a default value. I've seen a lot of code bases but never one which actually cared about its ABI compatibility. Can certainly be omitted, because students will invariably re-discover this use if it matters to them, but I included out of my own notion of completeness. Happy to remove if strong opinions exist

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md:35]: That could very well be. I'd be happy to move all the things to "basic". Admittedly, I put them under "advanced" in order to have _something_ under advanced.

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md:52]: Yes to part A, Not sure I understand your point about defining vs calling functions. Is there a suggested change to the wording there?

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md:82]: oh. nice idea!

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md:86]: Trying to get at something like this:
```
void foo(int x, int y);
void bar() { foo(5, 10); }
void bar2() { foo(7); } // not ok
void foo(int x, int y = 5);
void bar3() { foo(7); } // ok
void bar4() { foo(); } // not ok

void foo(int x=42, int y);
void bar5() { foo(5, 10); }
void bar6() { foo(7); }
void bar7() { foo(); }
```
https://godbolt.org/z/24cAzc

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [reply functions/defaulted_parameters.md:52]: > Not sure I understand your point about defining vs calling functions

I just meant that the lifetime of `std::string("hello world")` there doesn't matter at all to the _caller_ of the function, who may not even be aware that the default argument exists. If there's a bug in that code, it's a bug written-by and fixable-only-by the person who _defined_ the function with default arguments.  (This was relevant because at this point in the doc we're still in the "Basic" section that purports to have nothing to do with defining functions, only calling them. But probably that should be fixed by moving everything into "Basic".)

File: functions/defaulted_parameters.md
Comment (Quuxplusone) [reply functions/defaulted_parameters.md:86]: I might phrase that as "Subsequent redeclarations of the same function may add default argument values, which are then usable by callers; but a single argument cannot be given a default twice in the same translation unit."
And add: "A single argument can be given different defaults in different translation units."
And figure out some way to say "...but please never do that!" for both points.

File: functions/defaulted_parameters.md
Comment (vulder) [inline functions/defaulted_parameters.md:1]:
```suggestion
# Module name: defaulted parameters
```

File: functions/defaulted_parameters.md
Comment (vulder) [inline functions/defaulted_parameters.md:41]:
```suggestion
* Make calls to existing functions, passing parameters  [[Functions: calling functions]][1]
```
Where ever we can we should add references to other topics with links, so it is easier to navigate the different topics afterwards.
`[1]` refers to the link I suggested to add at the bottom of the file

File: functions/defaulted_parameters.md
Comment (vulder) [inline functions/defaulted_parameters.md:43]:
```suggestion
* Define member and non-member functions  [[Functions: member functions]][2]
```
@cjdb We should add `Functions: free functions` to P1725

File: functions/defaulted_parameters.md
Comment (vulder) [inline functions/defaulted_parameters.md:44]:
```suggestion
* Explain what a default constructor is and does  [[C++ object model: constructors]][3]
```

File: functions/defaulted_parameters.md
Comment (vulder) [inline functions/defaulted_parameters.md:87]:
```suggestion
[1]: ../functions/calling-functions.md
[2]: ../functions/member-functions.md
[3]: ../object-model/constructors.md
```

We should also add empty files with a small place holder for topics that are not yet created. See similar to the copy-semantics PR https://github.com/cplusplus/SG20/pull/55

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md:52]: Just gonna move it all to basic. :-)

Comment (rwdougla) [reply functions/defaulted_parameters.md]: done

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: updated, along with inclusion of previously "advanced" material

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: done

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: done

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: I updated free sections to 75 characters. Havent yet mucked around with list items. Still learning some of these features of markdown...

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: done

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: done

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: done (except for modification of P1725

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: done

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: done and done

File: functions/defaulted_parameters.md
Comment (rwdougla) [reply functions/defaulted_parameters.md]: Took a stab at it. Lemme know what you think of the new wording

File: functions/defaulted-parameters.md
Comment (Quuxplusone) [inline functions/defaulted-parameters.md:95]: `s/perhaps //`

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:18]:
```suggestion
    <td>Foundational</td>
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:22]:
```suggestion
    <td>Main</td>
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:26]:
```suggestion
    <td>Advanced</td>
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:41]:
```suggestion
## Foundational: Using and defining functions with default arguments
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:75]:
```suggestion
## Main: implementing *
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:90]:
```suggestion
## Advanced
```

**Comment by cjdb** [comment]
I'll approve this once Florian's changes have been made.

**Comment by zygoloid** [comment]
This language feature is called a "default argument", not a "default parameter" or a "defaulted parameter" or a "default argument value". We should presumably use the standard terminology and use it consistently.

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:39]:
```suggestion
Explain how default arguments work and how to define them.
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:56]:
```suggestion
1. Call to a function with a default argument with or without that argument specified
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:57]:
```suggestion
2. Declare a function with a default argument, and omit the default in the definition's signature
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:58]:
```suggestion
3. Explain what the requirements are for a type to be usable as a default argument
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:59]:
```suggestion
4. Explain when the lifetime of a defaulted argument begins and ends
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:67]:
```suggestion
* Calling an overloaded function with fewer arguments may be ambiguous with regard to an overload with default arguments
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:96]:
```suggestion
function different default arguments in different translation units.
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:8]:
```suggestion
of parameters, to allow a caller to avoid specifying arguments that 
```

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:47]:
```suggestion
* Make calls to existing functions, passing arguments [[Functions: calling functions]][1]
```

**Review by vulder — CHANGES_REQUESTED** [review]
@zygoloid 
Your right.Renaming this makes sense. This was probably introduced with porting [p1725](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1725r1.html) to gh issues.
I've update the issue #40 and added change suggestions to the doc. Could you please take a quick peak over the wording, specifically where we use argument/parameter?

@rwdougla Ok if we just switch 'parameter' with ' argument'? Could you also please rename the file, the gh api does not allow me to do this with a review suggestion.

File: functions/defaulted-parameters.md
Comment (zygoloid) [inline functions/defaulted-parameters.md:7]:
```suggestion
parameters. It may be of value to specify default arguments for some number 
```

File: functions/defaulted-parameters.md
Comment (zygoloid) [inline functions/defaulted-parameters.md:58]: This doesn't make sense to me: a type is not used as a default argument, an initializer is.

File: functions/defaulted-parameters.md
Comment (zygoloid) [inline functions/defaulted-parameters.md:59]:
```suggestion
4. Explain when the lifetime of a default argument begins and ends
```

File: functions/defaulted-parameters.md
Comment (zygoloid) [inline functions/defaulted-parameters.md]: This has never been true for inline functions, and is no longer true for any functions: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1766r1.html#default-argument-inconsistency

Comment (zygoloid) [reply functions/defaulted-parameters.md]: It is permissible to give functions different default arguments in different *scopes* (eg, a block-scope function declaration can have default arguments and those need not match those in any other scope), but that's an expert-level topic.

File: functions/defaulted-parameters.md
Comment (zygoloid) [inline functions/defaulted-parameters.md:94]:
```suggestion
Though a single parameter cannot be given a default argument twice in the same 
```

File: functions/defaulted-parameters.md
Comment (Quuxplusone) [reply functions/defaulted-parameters.md]: Off-topic for SG20, but: In what sense could the same lexical "scope" span two completely different "translation units"? I don't see any teachable difference between

```
// main.cpp
int foo() { extern int d(int x=1); d(); }
int bar() { extern int d(int x=2); d(); }
```
and
```
// foo.cpp
int d(int x=1); int foo() { d(); }
// bar.cpp
int d(int x=2); int foo() { d(); }
```
I agree that this is an "expert-level topic" and in fact teachers should be very clear with their students that you should _never ever_ give the same function different default arguments. (I simply teach that [default function arguments are the devil](https://quuxplusone.github.io/blog/2020/04/18/default-function-arguments-are-the-devil/), and leave it at that.)

Comment (rwdougla) [reply functions/defaulted-parameters.md]: I believe this is one of those topics that fall under my category of "Never write this, but you may need to be aware of it when dealing with other developers or long-lived codebases." IIUC, this is in the most "advanced" section we have in our skeleton.  Though now that it is forbidden across translation units, I'd be fine to strike the paragraph, altogether, unless there's something interesting there to keep.

File: functions/defaulted-parameters.md
Comment (rwdougla) [reply functions/defaulted-parameters.md:58]: Would such be better?
"Explain what the requirements are on a type for a value of that type to be usable as a default argument"

Comment (Quuxplusone) [reply functions/defaulted-parameters.md:58]: *Are* there any requirements on a type for a value of that type to be usable as a default argument? I can't think of any such requirements off the top of my head. You want a default argument of type `std::map<int, std::string>`?— no problem.

File: functions/defaulted-parameters.md
Comment (zygoloid) [reply functions/defaulted-parameters.md]: @Quuxplusone (Continuing off-topic discussion. Perhaps we should move this to email?)

> In what sense could the same lexical "scope" span two completely different "translation units"?

This can happen via `extern "C"`:
```
namespace A { extern "C" void f(int = 1); }
namespace B { extern "C" void f(int = 2); }
```
... declares the same function with different default arguments in two different scopes, both of which can span multiple translation units.

> I don't see any teachable difference between
> ```
> // main.cpp
> int foo() { extern int d(int x=1); d(); }
> int bar() { extern int d(int x=2); d(); }
> ```
> and
>```
> // foo.cpp
> int d(int x=1); int foo() { d(); }
> ```
> ```
> // bar.cpp
> int d(int x=2); int foo() { d(); }
> ```

Well, the former is valid and the latter is ill-formed (NDR). Whether that qualifies as a teachable difference would depend on the level of the audience, though even at expert level I think it's probably not worth mentioning. "Don't do that" seems like substantially better advice to me, both for declaring the same function with different default arguments, and for block-scope function declarations.

File: functions/defaulted-parameters.md
Comment (vulder) [inline functions/defaulted-parameters.md:1]:
```suggestion
# Functions: default argument
```
Fixes missing module name

File: functions/defaulted-parameters.md
Comment (vulder) [reply functions/defaulted-parameters.md:58]: What's the current status of this? Should we adapt the alternative wording from @rwdougla ?

**Comment by vulder** [comment]
I don't see many/any hard blocking issues here. Hence, we should discuss this PR at the next SG20 telecon.

**Review by vulder — APPROVED** [review]

File: functions/defaulted-parameters.md
Comment (rwdougla) [reply functions/defaulted-parameters.md]: I've removed the bullet point. I can't even recall what I had in mind for it, anyway. If someone has an idea as to why it would be useful, we can always add it back in later

**Pull request merged**
